# Debugging

During the pre-work you might have made an error or two before you succeeded in completing a challenge. Was that the case? How did you solve your problem? You definitely found a proper and better solution in the very end. But what do you do if you are struggling and can't find a good solution?

We can try and **debug** our code!

According to Wikipedia, "Debugging is the process of finding and resolving defects or problems within a computer program that prevent correct operation of computer software or a system". 

As you can probably see from the definition, the purpose of this process is to identify the mistake in your code to make it work as it should work. Debugging lets to identify the values of every variable on every step and trace errors in the code.

The important thing is to learn from mistakes you do while coding and avoid repeating them. 

Those, who did the technical interview, have met the problem of the Fibonacci sequence. The Fibonacci sequence is a sequence where each value in the sequence is the summed value of its two predecessors. Recall also that the sequence always starts with a 0 and then a 1. 

Let's solve this challenge!  

## Step-by-step solution

To reiterate, we want to generate a list of the first 100 Fibonacci numbers. Here, first we define the list with first 2 elements.

In [ ]:
fib=[0,1]


What should be done next? 

Now we need to calculate the value of next element.

In [ ]:
x=fib[0]
y=fib[1]
z=x+y
print(z)

1


Well, the value is calculated. But it is not in the list. Few seconds of googling gives us the necessary function:



In [ ]:
fib.append(z)

Well, the value is there. Let's check:

In [ ]:
print(fib) 

[0, 1, 1]


It seems that now we need to loop the sum 100 times. Then, on every iteration our code should calculate the sum, and append that value to the list. 



In [ ]:
fib = [0,1]
x = fib[0]
y = fib[1]
 
for i in range(100):
    z=x+y
    fib.append(z)

The work is done, no? Let's print the value of fib.


In [ ]:
print(fib)

[0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]


What is wrong here? We defined the values of x and y in the very beginning and they are the first and the second element of the list. However, *we need to change the values on every iteration!* 

In [ ]:
fib=[0,1]

for i in range(100):
    z=fib[i-1]+fib[i-2]
    fib.append(z)

In [ ]:
print(fib)

[0, 1, 1, 1, 1, 2, 2, 2, 3, 4, 4, 5, 7, 8, 9, 12, 15, 17, 21, 27, 32, 38, 48, 59, 70, 86, 107, 129, 156, 193, 236, 285, 349, 429, 521, 634, 778, 950, 1155, 1412, 1728, 2105, 2567, 3140, 3833, 4672, 5707, 6973, 8505, 10379, 12680, 15478, 18884, 23059, 28158, 34362, 41943, 51217, 62520, 76305, 93160, 113737, 138825, 169465, 206897, 252562, 308290, 376362, 459459, 560852, 684652, 835821, 1020311, 1245504, 1520473, 1856132, 2265815, 2765977, 3376605, 4121947, 5031792, 6142582, 7498552, 9153739, 11174374, 13641134, 16652291, 20328113, 24815508, 30293425, 36980404, 45143621, 55108933, 67273829, 82124025, 100252554, 122382762, 149397854, 182376579, 222635316, 271780616, 331774433]


On the first iteration we see that the values are taken more or less correct. *But on the second iteration everything fails!* Our Algorithm takes the same values.


Why?

As you can see, on the first iteration *it takes the negative values of the list, which are values taken from the end of the list.* 

<table>
<tbody>
<tr>
<td>Iteration</td>
<td>i</td>
<td>i-1</td>
<td>i-2</td>
<td>fib[i-1]</td>
<td>fib[i-2]</td>
<td>z</td>
</tr>
<tr>
<td>1</td>
<td>0</td>
<td>-1</td>
<td>-2</td>
<td>1</td>
<td>0</td>
<td>1</td>
</tr>
<tr>
<td>2</td>
<td>1</td>
<td>0</td>
<td>-1</td>
<td>0</td>
<td>1</td>
<td>1</td>
</tr>
<tr>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr>
<td>100</td>
<td>99</td>
<td>98</td>
<td>97</td>
<td>222635316</td>
<td>271780616</td>
<td>331774433</td>
</tr>
</tbody>
</table>

We can solve this issue as follows: we select only positive values by starting from i=2. 

In [ ]:
fib=[0,1]

for i in range(2,100):
    z=fib[i-1]+fib[i-2]
    fib.append(z)
    
print(fib)

[0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377, 610, 987, 1597, 2584, 4181, 6765, 10946, 17711, 28657, 46368, 75025, 121393, 196418, 317811, 514229, 832040, 1346269, 2178309, 3524578, 5702887, 9227465, 14930352, 24157817, 39088169, 63245986, 102334155, 165580141, 267914296, 433494437, 701408733, 1134903170, 1836311903, 2971215073, 4807526976, 7778742049, 12586269025, 20365011074, 32951280099, 53316291173, 86267571272, 139583862445, 225851433717, 365435296162, 591286729879, 956722026041, 1548008755920, 2504730781961, 4052739537881, 6557470319842, 10610209857723, 17167680177565, 27777890035288, 44945570212853, 72723460248141, 117669030460994, 190392490709135, 308061521170129, 498454011879264, 806515533049393, 1304969544928657, 2111485077978050, 3416454622906707, 5527939700884757, 8944394323791464, 14472334024676221, 23416728348467685, 37889062373143906, 61305790721611591, 99194853094755497, 160500643816367088, 259695496911122585, 420196140727489673, 679891637638612258, 110008777

## The Main Message

What did we do in the previous example? We were printing the values of every variable to see its value. We were checking the values of every variable on every iteration. And we basically tried to think as machine for a few minutes. 


What does debugging give us? Understanding the flow and reasons behind our mistakes. It explains you that machine is never wrong. Machine just follows your commands, which may be wrong.



Whenever you encounter a bug, try to understand what could go wrong by running different steps of your code separately. They should work on any inputs you provide and in any order you run your code. 